### PR TITLE
Update examples - use DataModule

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

partial update for #4694 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/__init__.py [hunk 1]
```diff
@@ -21,13 +22,26 @@
...
 from pytorch_lightning.utilities.parsing import AttributeDict, flatten_dict, is_picklable
 
-try:
-    from apex import amp
-except ImportError:
-    APEX_AVAILABLE = False
-else:
-    APEX_AVAILABLE = True
 
+def _module_available(module_path: str) -> bool:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/__init__.py:26]: nice, this is great. It's what I wanted to do for #2266 but never got around to it.
would be great to follow up on this and use this utility in other places too <3

File: pl_examples/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import os
+
+EXAMPLES_ROOT = os.path.dirname(__file__)
+PACKAGE_ROOT = os.path.dirname(EXAMPLES_ROOT)
+DATASETS_PATH = os.path.join(PACKAGE_ROOT, 'Datasets')
```
[the remark this answers is not in the document]
Comment (Borda) [reply pl_examples/__init__.py]: mean the first or second one?

File: pytorch_lightning/utilities/__init__.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/utilities/__init__.py:26]: yes, that is why I moved it in the end to PL util even it was developed and used for now in examples :]

Comment (tchaton) [reply pytorch_lightning/utilities/__init__.py:26]: Yes, let's use `_module_available` everywhere, it looks cleaner than trying to do imports.

File: pl_examples/basic_examples/backbone_image_classifier.py [hunk 3]
```diff
@@ -15,14 +15,16 @@
...
 
 import torch
-import pytorch_lightning as pl
 from torch.nn import functional as F
 from torch.utils.data import DataLoader, random_split
 
+import pytorch_lightning as pl
+from pl_examples import DATASETS_PATH
+
 try:
```
Comment (tchaton) [inline pl_examples/basic_examples/backbone_image_classifier.py:24]: Can you use HAS_TORCH_VISION there ?

**Review by tchaton — APPROVED** [review]
Great PR ! Let's use `_module_available` for all third party libraries. 
Would you mind adding a test for `_module_available` too ?

**Comment by Borda** [comment]
> Great PR ! Let's use `_module_available` for all third party libraries.
> Would you mind adding a test for `_module_available` too ?

it is already tested by its example - doctest

File: pl_examples/basic_examples/autoencoder.py [hunk 4]
```diff
@@ -13,17 +13,20 @@
...
 
-try:
+import pytorch_lightning as pl
+from pl_examples import TORCHVISION_AVAILABLE
+
+if TORCHVISION_AVAILABLE:
     from torchvision.datasets.mnist import MNIST
     from torchvision import transforms
-except ModuleNotFoundError:
+else:
```
Comment (SkafteNicki) [inline pl_examples/basic_examples/autoencoder.py:29]: What about `transforms` in the `else` case case? that would be undefined

Comment (Borda) [reply pl_examples/basic_examples/autoencoder.py:29]: I know that was a bit tricky, but we still have `torchvision` as requirements for examples...
basically, we can write a custom wrapper which does `toTensor()` or ignore this conversion

Comment (Borda) [reply pl_examples/basic_examples/autoencoder.py:29]: fixed :]

File: pl_examples/basic_examples/dali_image_classifier.py [hunk 5]
```diff
@@ -22,21 +22,21 @@
...
 
-try:
+if DALI_AVAILABLE:
     import nvidia.dali.ops as ops
-    import nvidia.dali.types as types
     from nvidia.dali.pipeline import Pipeline
     from nvidia.dali.plugin.pytorch import DALIClassificationIterator
-except (ImportError, ModuleNotFoundError):
+else:
     warn('NVIDIA DALI is not available')
```
Comment (SeanNaren) [inline pl_examples/basic_examples/dali_image_classifier.py:38]: I guess this isn't related to this PR, but shouldn't this be a hard crash if DALI isn't available?

**Review by SeanNaren — APPROVED** [review]

File: pl_examples/basic_examples/dali_image_classifier.py (hunk 5 above)
Comment (Borda) [reply pl_examples/basic_examples/dali_image_classifier.py:38]: what do you mean?

Comment (Borda) [reply pl_examples/basic_examples/dali_image_classifier.py:38]: ok, the case is that with tests running on CPU where we do not install DALI the testing will fail just because of this raising error, an alternative would be to return from this script if DALI is missing without crashing...

**Review by nateraw — APPROVED** [review]
LGTM 🚀

**Review by rohitgr7 — APPROVED** [review]
